### PR TITLE
Add cooldown to hypospray

### PR DESCRIPTION
**Pull request by Interrobang01**
## About the PR

Added a short (0.5 seconds) cooldown on the hypospray and its variants. This is due to a massive amount of arguing and support for the change in the discord.

Due to my lack of knowledge on C#, there are numerous issues in the code. Despite these, the code still works!
The issues are listed here:

- [X] Random dependancy
- [X] Variable definition in the middle of a bunch of checks
- [x] Check for delay is probably not in the ideal order
- [X] Nested if-statements to prevent C# from throwing an error (most likely very easy to fix but I don't know how)
- [x] Cooldown is started twice if hypo is used via hand interact
- [x] The bit that starts the delay probably isn't in the best spot
- [x] Actual cooldown numbers are not thought through well and likely need changing

Help fixing these is appreciated!

**Media**

https://user-images.githubusercontent.com/113810873/216882576-36e568c1-b4cc-4332-8861-6d05c297664e.mp4

- [X ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Hypospray Nerfer
- tweak: Hypospray, Hypopen, and Gorlax Hypospray now all have short cooldowns.

### CODE REVIEW

**Review by moonheart08 — COMMENTED** [review]
(obviously doublecheck this code. Treat it as a web edit, don't let github's commit suggestion button tempt you.)

**Comment by CaptainSqrBeard** [comment]
Isn't it breaks the essence of hypospray? I mean it should inject chemicals very fast.

It also doesn't help with nocturine+hypopen (or any other chem) meta as you have enough time to inject enough amount of chemical. It still just 2-3 clicks but you have to wait 0.5 sec between them. If you want so you can slip your target on soap and it will be unable to resist.

There should be other way to nerf this

**Comment by PJB3005** [comment]
Hypospray cooldown is fine. Hell on SS13 I think the cooldown is just like 1-2 seconds already due to how the character interaction logic works.

The important thing is that it's an instant action that has no do after, so you can apply it without the recipient standing still.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystemHypospray.cs [hunk 1]
```diff
@@ -65,6 +67,15 @@ public bool TryDoInject(EntityUid uid, EntityUid? target, EntityUid user, Hyposp
...
 
+            // Not a good location
+            var hasCooldownComponent = EntityManager.HasComponent<UseDelayComponent>(uid);
+
+            // Medipens and such use this system and don't have a delay, requiring extra checks
+            // This makes my eyes water, please advise
+            if (hasCooldownComponent == true)
+                if (_useDelay.ActiveDelay(uid, EntityManager.GetComponent<UseDelayComponent>(uid)))
+                    return false;
+
```
[the remark this answers is not in the document]
Comment (Interrobang01) [reply Content.Server/Chemistry/EntitySystems/ChemistrySystemHypospray.cs]: Needed UseDelayComponent to be made nullable, but it worked after that.

**Comment by GoodWheatley** [comment]
This change only nerfs players who are actually using the hypospray for healing people. The poisons antags can get access to are much stronger than the medicines that heal them, and a lethal dose of lexorin would only take 1 second to inject. But for a CMO who’s trying to treat a dozen injured patients, they’ll have to stand there for 3 seconds injecting bicaridine or dermaline before they could move onto the next patient. This nerf doesn’t reduce the lethality of hyposprays, it just makes their intended purpose of quickly healing people more aggravating to perform.

**Comment by EmoGarbage404** [comment]
This change is good. Injecting 20u in 2 seconds is still much faster than using a syringe or any other method. This also is a desperately needed balance tweak for antags. Like PJB said, the important factor is that theres no doafter for the injection itself.

**Comment by PJB3005** [comment]
@GoodWheatley look I haven't exactly played CMO in SS14 yet but I do have years of experience on SS13 and I genuinely fail to imagine a THE RIDE where you have enough patients lined up that instantly need the hypospray one after another that even a half-second cooldown is too much.

Also your scenario of "injecting a single patient for 3 seconds" doesn't even make any sense? You'd completely empty your hypospray on a single patient, so you'd need to be constantly refilling it and carrying bottles around. Using pills is a *much* saner solution in this use case.

> and a lethal dose of lexorin would only take 1 second to inject

A lethal dose of lexorin does not kill instantly, so what does this even matter? I don't know the ratios for antitox/charcoal but this really doesn't sound like it matters.

**Comment by PJB3005** [comment]
<details>
<summary>forbidden CMO tips</summary>

The hypospray is fucking trash at healing anyways, the only goo use is for filling it with chloral to keep shitters out of medbay
</details>

**Comment by GoodWheatley** [comment]
> 

I'll have you know I saved a dozen players with my hypospray and several bottles of dermaline after an atmos tech accidentally plasmaflooded and someone lit a welder in the hallway

**Comment by Interrobang01** [comment]
> If you think you can do it I'd suggest maybe adding a DoAfter for adding chems into the hypopen/hypospray, there's a current strategy of carrying around a deathmix-filled beaker to instantly reload the hypo with and moony makes a good point about the instant refill comboing with the instant injection

I agree with doing this, but it's a bit out of scope for this PR. I did see a suggestion for not loading the hypospray directly, but loading a hypospray cartridge with chems and putting that into the hypospray, just to slow things down.

**Comment by PJB3005** [comment]
I don't think a cooldown on refilling is that big of a deal. The real pain is inventory management time wasting, not 0.5 seconds of downtime.

Like I alluded to before, the hypospray isn't a cooler pill bottle. It's meant for quick administration of small amounts of  stabilizing agents (inaprov, dex+, ...) so that a patient stops getting worse, that's it. This utility is also what makes it a dangerous high-risk item that must be guarded with care by the CMO.

> I'll have you know I saved a dozen players with my hypospray and several bottles of dermaline after an atmos tech accidentally plasmaflooded and someone lit a welder in the hallway

The correct thing to do here that would've worked equally well was to inject every person with 5u inaprov or something so they stop dying in crit, then go over them 1-by-1 with a pill bottle of dermaline and then haul them to medbay if that isn't enough.

That's assuming our medbay balance isn't horrifically broken in the unending wait for bobby, of course, but I haven't played SS14 so.

**Comment by arealbeba** [comment]
hypospray cooldown is cool, I'm tired of those powergamers injecting tons of chemicals in just a few seconds

**Comment by Interrobang01** [comment]
Gonna try to finish this tonight. I'll check off the "is the cooldown balanced" part because if it's not balanced we can just change it later. I'll ask on the discord to try and fix the rest of the issues.

**Comment by Interrobang01** [comment]
turns out there's a check in the BeginDelay function that halts it if a delay is already in progress, so i'll just check that one off

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**